internal/api/routes.go

```diff
@@ -9,19 +9,30 @@ import (
 	"github.com/edulinq/autograder/internal/api/logs"
 	"github.com/edulinq/autograder/internal/api/static"
 	"github.com/edulinq/autograder/internal/api/users"
+	"github.com/edulinq/autograder/internal/util"
 )
 
-var baseRoutes = []*core.Route{
+var baseRoutes = []core.Route{
 	core.NewRedirect("GET", ``, `/static/index.html`),
 	core.NewRedirect("GET", `/`, `/static/index.html`),
 	core.NewRedirect("GET", `/index.html`, `/static/index.html`),
 
-	core.NewRoute("GET", `/static`, static.Handle),
-	core.NewRoute("GET", `/static/.*`, static.Handle),
+	core.NewBaseRoute("GET", `/static`, static.Handle),
+	core.NewBaseRoute("GET", `/static/.*`, static.Handle),
 }
 
-func GetRoutes() *[]*core.Route {
-	routes := make([]*core.Route, 0)
+type APIDescription struct {
+	Endpoints map[string]map[string]EndpointDescription
```

Comment:
Why the double map?
You have a type called endpoints which in JSON will have a single field called "Endpoints",
which is a map with a single field called "endpoints".
You can see how rediculious this is getting.

Just make it a single map `{basePath: description, ...}`.